Tests/UnitTests/Tests/AsyncBlockTests.cpp

```diff
@@ -987,4 +999,82 @@ DEFINE_TEST_CLASS(AsyncBlockTests)
         VERIFY_SUCCEEDED(XAsyncGetStatus(&async, true));
         VERIFY_ARE_EQUAL((DWORD)WAIT_OBJECT_0, WaitForSingleObject(context.evt, 2500));
     }
+
+    DEFINE_TEST_CASE(VerifyBeginAfterTerminate)
+    {
+        XAsyncBlock async{};
+        VERIFY_SUCCEEDED(XTaskQueueCreate(XTaskQueueDispatchMode::ThreadPool, XTaskQueueDispatchMode::ThreadPool, &async.queue));
+
+        async.callback = [](XAsyncBlock* async)
+        {
+            (*(bool*)async->context) = true;
+        };
+
+        auto provider = [](XAsyncOp op, const XAsyncProviderData* data)
+        {
+            switch(op)
+            {
+                case XAsyncOp::Begin:
+                    return XAsyncSchedule(data->async, 0);
+
+                case XAsyncOp::Cleanup:
+                    break;
+
+                default:
+                    VERIFY_FAIL();
+                    break;
+            }
+            return S_OK;
+        };
+
+        // Terminate the queue
+        XTaskQueueTerminate(async.queue, true, nullptr, nullptr);
+
+        bool callbackCalled = false;
+        async.context = &callbackCalled;
+
+        // XAsyncBegin should succeed.  The callback should have been called anyway, even
```

Comment:
does that mean we're potentially calling callback:
- from inside the XAsyncBegin callstack
- on the wrong thread

Both of these scare me, because now I have to write code to account for them happening all the time :(

The first one is especially bad if I have non recursive locks I'm taking

Reply:
Yes, it does.  The alternative, if we hit this case, is that we never call your callback and your task never completes.

Reply:
After much discussion we decided not to do this but to instead early check for termination and, if we later on get surprised with failing to submit a completion, ,we failfast the process.